### DISCUSSION TITLE
SSL Solved

### DISCUSSION THREAD

**Discussion by MPatriot** [oqtane/oqtane.framework#1983] [General]
For those new to Oqtane, as am I, one of the first questions I had was how do I get my development environment to automatically start using https: instead of http: like it does in a regular Blazor application? (_If when creating you chose the 'Configure for Https' option_)

This is absolutely critical for me in development because I use Google as my SMTP host(paid) and cannot test locally without https and it **also means Scheduled Jobs like Notifications (using SMTP) cannot run on local host**.

I searched the Issues and Discussions here and found very little information regarding this simple issue other than some links to how to set up ssl and that Oqtane is already set up for ssl.  So sure, SMTP works fine on my server that is setup with certificates and everything but not on localhost.

Anyway, I put the issue aside knowing it all worked on the server and decided to get back to it later.  This morning I decided to get back to it and here is how I solved it:  (**Caveat:** this violates the guidelines of changing Oqtane code and might be something you have to redo if / when a newer version of Oqtane is published.)

1. **Open Oqtane.sln** in Visual Studio
1. Using Visual Studio Solution Explorer Go to Oqtane.Server>Properties and **copy / paste launchSettings.json** (I renamed the copy to launchSettings - Orig.json)
2. The first section in the file should look like this:
"iisSettings": {
    "windowsAuthentication": false,
    "anonymousAuthentication": true,
    "iisExpress": {
      "applicationUrl": "http://localhost:44357/",
      "sslPort": 0
    }
Notice applicationUrl port is 44357 and sslPort is 0 - **leave it like that for now** but pick a sslPort number, I picked 44358)

3. **Run the Oqtane.Server project**, Click the 'Gear' icon and then Admin Dashboard button.  Choose Site Settings and in the **Alias** box there should already be 'localhost:44357'  add a comma(no spaces) after that and any other aliases you may have added and then **add 'localhost:44358'** (or whatever port you picked.)  Save and close running application.

4. Back in Visual Studio **edit the launchSettings.json**:
Change the sslPort to 44358 or whatever port you picked...this example will just show the port I picked for ssl
 "iisSettings": {
    "windowsAuthentication": false,
    "anonymousAuthentication": true,
    "iisExpress": {
      "applicationUrl": "http://localhost:44357/",
      "**sslPort": 44358**
    }
Go down to the bottom and **find the "Oqtane.Server" entry**
   "Oqtane.Server": {
      "commandName": "Project",
      "launchBrowser": true,
       "inspectUri": "{wsProtocol}://{url.hostname}:{url.port}/_framework/debug/ws-proxy?browser={browserInspectUri}",
      "environmentVariables": {
        "ASPNETCORE_ENVIRONMENT": "Development"
      },
      "applicationUrl": "http://localhost:44357/"
    }
**Change the Oqtane.Server entry to:**
    "Oqtane.Server": {
      "commandName": "Project",
      "launchBrowser": true,
      "inspectUri": "{wsProtocol}://{url.hostname}:{url.port}/_framework/debug/ws-proxy?browser={browserInspectUri}",
      "environmentVariables": {
        "ASPNETCORE_ENVIRONMENT": "Development"
      },
      "**applicationUrl": "https://localhost:44358;http://localhost:44357**"
    } 
**Save the launchSettings.json file** and run the application again...
**WaLa!**  It should now be running with https://localhost:44358/ url in the browser!

Cheers!
- Mark

**Comment by sbwalker** [comment]
I generally just use GMail for my development environment SMTP. You simply need to specify the Server Name: smtp.gmail.com:465 and use Basic Authentication and Enable SSL (and provide your GMail account details). Hotmail can be used in a similar manner. I personally find this to be easier than configuring a local environment.

**Reply by MPatriot** [reply]
Yes, I agree, its definitely easier for many reasons, though this isn't GMail, its Google Mail - smtp.googlemail.com:587
It's like less than $7 bucks a month per address.  Each address can have 30 aliases which easily covers everything
mark@domain.com, support@domain.com, no-reply@domain.com, contact@domain.com, notifications@domain.com ect ect
You can even use aliases for other domains you control ie... mark@domain.com - mark@domain2.com but I don't do that, to messy.
Its funny, I always choose or code EnableSSL but it seems to hate http vs https.  It's no problem to even connect with a console app or  any other app but it seems to hate browsers sending requests from http no mater what you tell it.

The main reason I use Google Mail is my emails make it to the inbox, never spam or even worse ... lost in a black hole, not even denied, just gone!  I don't remember when but in the late 1990's or early 2000's sending emails from web servers stopped working.  I would sent test emails from web servers to my personal accounts (at the time a comcast) and would never get them.  I even called Comcast about it and they just played dumb, "You aren't on any black lists or blocked at all"
I even follow a guy who uses Builderall and Kartra.  All of his Kartra mails get to my inbox, sometimes promotions depending on his subject line (Words like "Trial" or "Free"), but all of his Builderall emails go straight to spam.  I find a lot of gmails in spam also.
Google never messes with my emails, I often test them to myself and depending on the subject line, some go to the promotions folder.  None ever go to spam because I am very careful not send unsolicited emails (technically not spam but people often mark them as such) and always have a proper footer with unsubscribe links(if applicable) sender location / contact info ect.  It's possible that Builderall emails get sent to spam because another guy on their shared email server IS spaming, so even if everything right on your part, the guy next to you is messing it up for all.

Good stuff, might be way of topic now:)
-Mark
